go/nbs/s3_table_persister.go

```diff
@@ -184,3 +194,180 @@ func (s partsByPartNum) Less(i, j int) bool {
+	sent, failed, done := make(chan s3UploadedPart), make(chan error), make(chan struct{})
+	var uploadWg sync.WaitGroup
+	sendPart := func(partNum int64, doUpload uploadFn) {
+		defer uploadWg.Done()
+
+		// Check if upload has been terminated
+		select {
+		case <-done:
+			return
+		default:
+		}
+
+		etag, err := doUpload()
+		if err != nil {
+			failed <- err
+			return
+		}
+		// Try to send along part info. In the case that the upload was aborted, reading from done allows this worker to exit correctly.
+		select {
+		case sent <- s3UploadedPart{int64(partNum), etag}:
+		case <-done:
+			return
+		}
+	}
+
+	numCopyParts := len(copies)
+	numManualParts := getNumParts(len(buff), s3p.partSize) // TODO: What if this is too big?
+	numParts := numCopyParts + numManualParts
+	uploadWg.Add(numParts)
+	for i, cp := range copies {
+		partNum := int64(i + 1) // parts are 1-indexed
+		go sendPart(partNum, func() (etag string, err error) {
+			return s3p.uploadPartCopy(cp.name, cp.dataLen, key, uploadID, partNum)
+		})
+	}
+	for i := numCopyParts + 1; i <= numParts; i++ {
+		start, end := (i-numCopyParts-1)*s3p.partSize, (i-numCopyParts)*s3p.partSize
+		if i == numParts { // If this is the last part, make sure it includes any overflow
+			end = len(buff)
+		}
+		partNum := int64(i)
+		go sendPart(partNum, func() (etag string, err error) {
+			return s3p.uploadPart(buff[start:end], key, uploadID, partNum)
+		})
+	}
+
+	go func() {
+		uploadWg.Wait()
+		close(sent)
+		close(failed)
+	}()
+
+	multipartUpload := &s3.CompletedMultipartUpload{}
+	var firstFailure error
+	for cont := true; cont; {
+		select {
+		case sentPart, open := <-sent:
+			if open {
+				multipartUpload.Parts = append(multipartUpload.Parts, &s3.CompletedPart{
+					ETag:       aws.String(sentPart.etag),
+					PartNumber: aws.Int64(sentPart.idx),
+				})
+			}
+			cont = open
+
+		case err := <-failed:
+			if err != nil && firstFailure == nil { // nil err may happen when failed gets closed
+				firstFailure = err
+				close(done)
+			}
+		}
+	}
+
+	if firstFailure == nil {
+		close(done)
+	}
+	sort.Sort(partsByPartNum(multipartUpload.Parts))
+	return multipartUpload, firstFailure
+}
+
+type copyPart struct {
+	name    string
+	dataLen int64
+}
+
+type manualPart struct {
+	r                  io.Reader
+	buffStart, buffEnd int64
+}
+
+func dividePlan(sources chunkSources, plan compactionPlan) (copies []copyPart, manuals []manualPart, buff []byte) {
+	buffSize := uint64(len(plan.mergedIndex))
+	var offset int64
+	for i, src := range sources {
+		dataLen := plan.chunkDataLens[i]
+		if dataLen >= defaultS3PartSize { // Big enough to copy part!
+			copies = append(copies, copyPart{src.hash().String(), int64(dataLen)})
+		} else {
+			manuals = append(manuals, manualPart{src.reader(), offset, offset + int64(dataLen)})
+			offset += int64(dataLen)
+			buffSize += dataLen
+		}
+	}
+	buff = make([]byte, buffSize)
+	copy(buff[buffSize-uint64(len(plan.mergedIndex)):], plan.mergedIndex)
```

Comment:
index can be pretty large. why make a copy? i think the caller already owns the plan.